### PR TITLE
feature: convert design-system to typescript - part 1

### PR DESCRIPTION
**Pull request by DDDDDanica**
## Explanation
**New enums introduced for design-system**
- Color
- TypographyType
- TextType
- Size
- BorderStyle
- BorderRadius
- AlignItems
- JustifyContent

**Modify the old constants**
- `BACKGROUND_COLORS`
- `BORDER_COLORS`
- `TEXT_COLORS`
- `ICON_COLORS`

**Update some unit tests to remove warnings**

Typeforce some usages of colors, eg: Icon has to use `ICON_COLORS`

## Pre-merge author checklist

- [ ] I've clearly explained:
  - [ ] What problem this PR is solving
  - [ ] How this problem was solved
  - [ ] How reviewers can test my changes
- [ ] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

### CODE REVIEW

**Comment by brad-decker** [comment]
@DDDDDanica I know its going to cause a lot of extra work but I think for reviewability it would be best to break this up.

Step 1: Convert design-system to TS without changing enums.
Steps 2-*: One old Object map at a time, separate PR for each. This gives us way more auditability in the future too instead of one commit with all these changes in it.

**Comment by brad-decker** [comment]
@georgewrmarshall @garrettbear 😬 Winter (typescript) is coming.

**Comment by brad-decker** [comment]
Oh I see you already did phase 1. so just the individual constant changes then

File: ui/helpers/constants/design-system.ts [hunk 1]
```diff
@@ -338,7 +331,7 @@ export const RESIZE = {
   HORIZONTAL: 'horizontal',
   VERTICAL: 'vertical',
   INITIAL: 'initial',
-  INHERIT: 'inherit',
+  inherit: 'inherit',
```
Comment (DDDDDanica) [inline ui/helpers/constants/design-system.ts:334]: Revert back

File: ui/components/component-library/avatar-account/avatar-account.stories.js [hunk 2]
```diff
@@ -1,9 +1,9 @@
 import React from 'react';
 import Box from '../../ui/box/box';
 import {
-  ALIGN_ITEMS,
+  AlignItems,
   DISPLAY,
-  SIZES,
+  Size as SizeType,
```
Comment (georgewrmarshall) [inline ui/components/component-library/avatar-account/avatar-account.stories.js:6]: Instead of using aliases here I think it's probably better to update the names of the story and use the `storyName` to change the label. The `DefaultStory` is a good example. The name of the story is then changed to `DefaultStory.storyName = 'Default';`

```
export const SizeStory = (args) => (
...
);

SizeStory.storyName = 'Size'
```

**Review by georgewrmarshall — COMMENTED** [review]
I still need to finish a fine tooth comb through of all the files but so far it looks good. Still need to discuss why we can't just convert all design system types to enums as I think that would help with consistency. Left one comment regarding story naming but we could cover that on the design system team side so not blocking

File: ui/components/component-library/avatar-account/avatar-account.stories.js (hunk 2 above)
Comment (georgewrmarshall) [reply ui/components/component-library/avatar-account/avatar-account.stories.js:6]: This is non blocking and we can update but it's more valuable to document with the default enum names in our component stories and docs

Comment (DDDDDanica) [reply ui/components/component-library/avatar-account/avatar-account.stories.js]: Thanks for the tip. it is an easy fix with limited files to i followed your suggestion, now it's in the codebase :)

File: ui/contexts/metametrics.js [hunk 3]
```diff
@@ -87,6 +87,7 @@ export function MetaMetricsProvider({ children }) {
         options,
       );
     },
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (brad-decker) [inline ui/contexts/metametrics.js:90]: Remove this comment and allow the warning to persist.

File: ui/css/itcss/README.md [hunk 4]
```diff
@@ -1,2 +1,3 @@
 ### Deprecated
+
```
Comment (brad-decker) [inline ui/css/itcss/README.md:2]: Looks like an autoformat change that isn't part of this change.

File: ui/helpers/constants/design-system.ts [hunk 5]
```diff
@@ -1,236 +1,227 @@
...
-  XS: 'xs',
-  SM: 'sm',
-  MD: 'md',
-  LG: 'lg',
-  XL: 'xl',
-  INHERIT: 'inherit', // Used for Text, Icon, and Button components to inherit the parent elements font-size
-  AUTO: 'auto',
-  NONE,
-};
+export enum TextType {
```
Comment (brad-decker) [inline ui/helpers/constants/design-system.ts:164]: TextType should probably be TextVariant

File: ui/helpers/constants/design-system.ts [hunk 6]
```diff
@@ -1,236 +1,227 @@
...
+}
 
-export const ALIGN_ITEMS = {
-  FLEX_START,
-  FLEX_END,
-  CENTER,
-  BASELINE: 'baseline',
-  STRETCH: 'stretch',
-};
+export enum AlignItems {
```
Comment (brad-decker) [inline ui/helpers/constants/design-system.ts:209]: Add a comment above this to explain our stance on this pluralized enum, considering that align-items is the css prop and that this will align *all* items not one.

File: ui/pages/unlock-page/README.mdx [hunk 7]
```diff
@@ -21,7 +21,7 @@ ArgsTable doesn't work with SectionShape
 | -------------------------- | -------------------------------------------------------------------------- |
 | `history`                  | History router for redirect after action `object`                          |
 | `isUnlocked`               | If isUnlocked is true will redirect to most recent route in history `bool` |
-| `onRestore`                | onClick handler for "Forgot password?" link `func`      |
+| `onRestore`                | onClick handler for "Forgot password?" link `func`                         |
```
Comment (brad-decker) [inline ui/pages/unlock-page/README.mdx:24]: This change is a good change but likely needs extracted from this PR.

File: ui/pages/send/send-footer/send-footer.test.js [hunk 8]
```diff
@@ -42,7 +42,7 @@ describe('SendFooter Component', () => {
 
   it('should match snapshot', () => {
     const { container } = renderWithProvider(
-      <SendFooter {...props} />,
+      <SendFooter {...props} disabled={false} />,
```
Comment (brad-decker) [inline ui/pages/send/send-footer/send-footer.test.js:45]: We should extract these changes to a different PR to cleanup our tests. Good work but we don't want to hide these changes in the git history, or confuse the pr review by trying to figure out why these changes were necessary.

File: ui/components/app/asset-list/asset-list.test.js [hunk 9]
```diff
@@ -11,7 +11,8 @@ const render = () => {
       ...mockState.metamask,
     },
   });
-  return renderWithProvider(<AssetList />, store);
+  // eslint-disable-next-line no-empty-function
+  return renderWithProvider(<AssetList onClickAsset={() => {}} />, store);
```
Comment (pedronfigueiredo) [inline ui/components/app/asset-list/asset-list.test.js:15]: Can you revert this change and make it in another PR for better audibility?

File: ui/components/app/create-new-vault/create-new-vault.test.js [hunk 10]
```diff
@@ -13,13 +13,16 @@ const store = configureStore({
 
 describe('CreateNewVault', () => {
   it('renders CreateNewVault component and shows Secret Recovery Phrase text', () => {
-    renderWithProvider(<CreateNewVault submitText="Import" />, store);
+    renderWithProvider(
+      <CreateNewVault submitText="Import" onSubmit={jest.fn()} />,
```
Comment (pedronfigueiredo) [inline ui/components/app/create-new-vault/create-new-vault.test.js:17]: Can you revert this change on this PR and make it separately?

File: ui/pages/notifications/notifications.js [hunk 11]
```diff
@@ -130,13 +130,13 @@ export default function Notifications() {
 }
 
 NotificationItem.propTypes = {
-  notification: {
+  notification: PropTypes.exact({
```
Comment (brad-decker) [inline ui/pages/notifications/notifications.js:133]: Is this change required for the work of making the design-system enums?

File: ui/components/component-library/avatar-favicon/avatar-favicon.test.js [hunk 12]
```diff
@@ -26,7 +27,7 @@ describe('AvatarFavicon', () => {
 
   it('should render fallback image if no ImageSource is provided', () => {
     const { container } = render(
-      <AvatarFavicon data-testid="avatar-favicon" />,
+      <AvatarFavicon data-testid="avatar-favicon" name="icon" />,
```
Comment (pedronfigueiredo) [inline ui/components/component-library/avatar-favicon/avatar-favicon.test.js:30]: Can you revert this change on this PR and make it separately?

File: ui/components/component-library/picker-network/picker-network.test.js [hunk 13]
```diff
@@ -57,7 +57,11 @@ describe('PickerNetwork', () => {
   // className
   it('should render with custom className', () => {
     const { getByTestId } = render(
-      <PickerNetwork data-testid="picker-network" className="test-class" />,
+      <PickerNetwork
+        data-testid="picker-network"
+        className="test-class"
+        label="custom"
```
Comment (pedronfigueiredo) [inline ui/components/component-library/picker-network/picker-network.test.js:63]: Can you revert this change on this PR and make it separately?

**Review by brad-decker — CHANGES_REQUESTED** [review]
There are some test clean up in here, a couple eslint rule overrides that shouldn't be overidden (we must fix errors, but warnings are helpful to track what should be changed), and some markdown formatting that is likely unrelated to this work. IF we could separate those changes out into separate prs I know the engineer community would be grateful!

File: ui/components/component-library/text/text.test.js [hunk 14]
```diff
@@ -110,19 +110,18 @@ describe('Text', () => {
   it('should render the Text with proper text color class name', () => {
     const { getByText } = render(
       <>
-        <Text color={COLORS.TEXT_DEFAULT}>text-default</Text>
-        <Text color={COLORS.TEXT_ALTERNATIVE}>text-alternative</Text>
-        <Text color={COLORS.TEXT_MUTED}>text-muted</Text>
-        <Text color={COLORS.OVERLAY_INVERSE}>overlay-inverse</Text>
```
Comment (pedronfigueiredo) [inline ui/components/component-library/text/text.test.js:116]: Any reason to remove this color and respective test?

File: ui/components/ui/account-list/account-list.test.js [hunk 15]
```diff
@@ -38,6 +38,10 @@ const render = () => {
     },
     allAreSelected: () => true,
     nativeCurrency: 'USD',
+    selectNewAccountViaModal: jest.fn(),
```
Comment (pedronfigueiredo) [inline ui/components/ui/account-list/account-list.test.js:41]: Can you revert this change on this PR and make it separately?

File: ui/components/ui/actionable-message/actionable-message.js [hunk 16]
```diff
@@ -49,6 +49,7 @@ export default function ActionableMessage({
         clearTimeout(timeout);
       };
     },
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (pedronfigueiredo) [inline ui/components/ui/actionable-message/actionable-message.js:52]: Can you revert this change on this PR and make it separately?

File: ui/components/ui/box/box.stories.js [hunk 17]
```diff
@@ -288,254 +288,265 @@ export const Margin = (args) => {
...
-        <Typography color={COLORS.TEXT_DEFAULT}>
-          COLORS.WARNING_DEFAULT
+        <Typography color={TextColor.textDefault}>
+          BorderColor.warningDefault
         </Typography>
       </Box>
     </>
   );
 };
+BorderColorStory.storyName = 'BorderColor';
```
Comment (pedronfigueiredo) [inline ui/components/ui/box/box.stories.js:547]: Can you revert this change on this PR and make it separately?

File: ui/components/ui/box/box.stories.js [hunk 18]
```diff
@@ -545,84 +556,85 @@ export const BorderRadius = () => {
...
+        alignItems={AlignItems.center}
         style={{ height: '250px', width: '250px' }}
       >
-        BORDER_RADIUS.FULL 50%
+        BorderRadius.full 50%
       </Box>
     </>
   );
 };
+BorderRadiusStories.storyName = 'BorderRadius';
```
Comment (pedronfigueiredo) [inline ui/components/ui/box/box.stories.js:637]: Can you revert this change on this PR and make it separately?``

File: ui/components/ui/box/box.stories.js (hunk 17 above)
Comment (DDDDDanica) [reply ui/components/ui/box/box.stories.js:547]: Ah this is actually needed, because the storyname `BorderColor` has the same name as variable`BorderColor.borderDefault`, so i changed the storyname :)

File: ui/components/component-library/text/text.test.js (hunk 14 above)
Comment (DDDDDanica) [reply ui/components/component-library/text/text.test.js:116]: There was no TextColor under this key, i will add it back for transparency 🙏

File: ui/components/component-library/picker-network/picker-network.test.js (hunk 13 above)
Comment (DDDDDanica) [reply ui/components/component-library/picker-network/picker-network.test.js:63]: Will do !

File: ui/components/ui/box/box.stories.js (hunk 18 above)
Comment (DDDDDanica) [reply ui/components/ui/box/box.stories.js:637]: Actually It is needed to avoid name conflict ~
https://github.com/MetaMask/metamask-extension/pull/17518#discussion_r1093507043

File: ui/pages/notifications/notifications.js (hunk 11 above)
Comment (DDDDDanica) [reply ui/pages/notifications/notifications.js]: reverted 🙏

**Review by brad-decker — DISMISSED** [review]

**Review by garrettbear — DISMISSED** [review]

**Review by brad-decker — DISMISSED** [review]

**Review by brad-decker — APPROVED** [review]

**Review by garrettbear — APPROVED** [review]
